src/main/java/sg/ncl/DataController.java

```diff
@@ -0,0 +1,236 @@
+package sg.ncl;
+
+import lombok.extern.slf4j.Slf4j;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.validation.BindingResult;
+import org.springframework.validation.FieldError;
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.annotation.ModelAttribute;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.servlet.mvc.support.RedirectAttributes;
+import sg.ncl.domain.ExceptionState;
+import sg.ncl.exceptions.WebServiceRuntimeException;
+import sg.ncl.testbed_interface.DataResource;
+import sg.ncl.testbed_interface.Dataset;
+import sg.ncl.testbed_interface.RestUtil;
+
+import javax.servlet.http.HttpSession;
+import javax.validation.Valid;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Optional;
+
+import static sg.ncl.domain.ExceptionState.FORBIDDEN_EXCEPTION;
+
+/**
+ * Created by dcsjnh on 11/17/2016.
+ */
+@Controller
+@RequestMapping("/data")
+@Slf4j
+public class DataController extends MainController {
+
+    private static final String CONTRIBUTE_DATA_PAGE = "data_contribute";
+
+    @RequestMapping
+    public String data(Model model, HttpSession session) throws Exception {
+        DatasetManager datasetManager = new DatasetManager();
+
+        HttpEntity<String> request = createHttpEntityHeaderOnly();
+        ResponseEntity response = restTemplate.exchange(properties.getData(), HttpMethod.GET, request, String.class);
+        String dataResponseBody = response.getBody().toString();
+
+        JSONArray dataJsonArray = new JSONArray(dataResponseBody);
+        for (int i = 0; i < dataJsonArray.length(); i++) {
+            JSONObject dataInfoObject = dataJsonArray.getJSONObject(i);
+            Dataset dataset = extractDataInfo(dataInfoObject.toString());
+            datasetManager.addDataset(dataset);
+        }
+
+        model.addAttribute("allDataMap", datasetManager.getDatasetMap());
+        return "data";
+    }
+
+    @RequestMapping(value={"/contribute", "/contribute/{id}"}, method= RequestMethod.GET)
+    public String contributeData(Model model, @PathVariable Optional<String> id) throws Exception {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS00112)